keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2168 @@
+progress is lost. The NodeMaintenance implementation should also utilize existing node's
+`.spec.unschedulable` field, which prevents new pods from being scheduled on such a node.
+
+We will deprecate the `kubectl drain` as the main mechanism for draining nodes and drive the whole
+process via a declarative API. This API can be used either manually or programmatically by other
+drain implementations (e.g. cluster autoscalers). 
+
+To support workload migration, a new controller should be introduced to observe the NodeMaintenance
+objects and then select pods for evacuation. The pods should be selected by node (`nodeSelector`)
+and the pods should be gradually evacuated according to the workload they are running.
+Controllers can then implement the migration/termination either by reacting to the Evacuation API
+or by reacting to the NodeMaintenance API if they need more details.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster admin I want to have a simple interface to initiate a node drain/maintenance without
+any required manual interventions. I want to have an ability to manually switch between the
+maintenance phases (Planning, Cordon, Drain, Drain Complete, Maintenance Complete). I also want to
+observe the node drain via the API and check on its progress. I also want to be able to discover
+workloads that are blocking the node drain.
+
+#### Story 2
+
+As an application owner, I want to run single replica applications without disruptions and have the
+ability to easily migrate the workload pods from one node to another. This also applies to
+applications with larger number of replicas that prefer to surge (upscale) pods first rather than
+downscale.
+
+#### Story 3
+
+Cluster or node autoscalers that take on the role of `kubectl drain` want to signal the intent to
+drain a node using the same API and provide a similar experience to the CLI counterpart.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+- This KEP depends on [Evacuation API KEP](https://github.com/kubernetes/enhancements/issues/4563).
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+A misconfigured .spec.nodeSelector could select all the nodes (or just all master nodes) in the
+cluster. This can cause the cluster to get into a degraded and unrecoverable state.
+
+An admission plugin ([NodeMaintenance Admission](#nodemaintenance-admission)) is introduced to
```

Comment:
But this will be unrecoverable, right? if the master nodes are not available it means will you not be able to connect to the apiserver

Reply:
The admission restriction wouldn't prevent you marking all the nodes as under maintenance. It'd just warn you that this was a bad idea.
(I'd prefer to have a ValidatingAdmissionPolicy rather than code, but we don't have lifecycle management for this kind of thing. In other words, kubeadm can't deploy recommended ValidatingAdmissionPolicies for us, so we shouldn't assume they are in place.)

- You can mark all the nodes as being covered by maintenance (and this may well be true!)
- If you as a cluster adminstrator then go and pull out the power cords or whatever, you can try to recover by plugging them back in.
- Static Pods don't stop during a node drain, right? So even if you do mark every node as under maintenance, you may still have a control plane. Perhaps, no working cluster network and no DNS, but you still have a control plane.

Reply:
Yes, it can still be a valid strategy. Assuming your control plane is deployed correctly - basically that it shuts down last or runs on the side (invisible to kube).

Reply:
👀 trying to better understand impacts of static Pods drain in the kubeadm scenario

Reply:
As I mentioned in https://github.com/kubernetes/enhancements/pull/4213#discussion_r1781840501, you can control which component goes down first and last and use it for gradual upgrades in clusters. There would have to be a specific configuration for a kubeadm cluster to let it downscale nicely. Depending on the cluster setup (HA, load balancing).

Reply:
The problem is that you can't really use the loadbalancer for the communication between kubelet and kube-apiserver.

The Kubernetes skew policy states: "kubelet must not be newer than kube-apiserver.". This means that if you configure kubelet to use a loadbalancer, during upgrades you would violate the Kubernetes skew policy (example, it can happen that a 1.31 kubelet communicates with a 1.30 kube-apiserver through the loadbalancer). That's why kubeadm started configuring kubelet to use the local kube-apiserver instead.

Reply:
Maybe we should consider solving the static pods in a different proposal or a follow-up? It might be worthwhile to take an incremental approach to a problem that spans across many SIGs and has a potential of wide impact on the ecosystem and current uses

Reply:
I like the idea of a PR against the same KEP ID that adds in treatment of static Pods.

`kubectl drain` today doesn't evict static Pods, and replacements would schedule&dagger; to the same node anyway.

Reply:
&dagger; _no actual scheduling_